src/realm/parser/query_builder.cpp

```diff
@@ -0,0 +1,970 @@
+                                                       value_of_type_for_query<String>(expr.table_getter, rhs, args));
+            break;
+        case type_Binary:
+            add_binary_constraint_to_query(query, cmp.op, value_of_type_for_query<Binary>(expr.table_getter, lhs, args),
+                                                          value_of_type_for_query<Binary>(expr.table_getter, rhs, args));
+            break;
+        case type_Link:
+            add_link_constraint_to_query(query, cmp.op, expr, link_argument(lhs, rhs, args));
+            break;
+        default:
+            throw std::logic_error(util::format("Object type '%1' not supported", data_type_to_str(expr.col_type)));
+    }
+}
+
+template<typename T>
+void do_add_null_comparison_to_query(Query &query, Predicate::Operator op, const PropertyExpression &expr)
+{
+    Columns<T> column = expr.table_getter()->template column<T>(expr.col_ndx);
+    switch (op) {
+        case Predicate::Operator::NotEqual:
+            query.and_query(column != realm::null());
+            break;
+        case Predicate::Operator::Equal:
+            query.and_query(column == realm::null());
+            break;
+        default:
+            throw std::logic_error("Only 'equal' and 'not equal' operators supported when comparing against 'null'.");
+    }
+}
+
+template<>
+void do_add_null_comparison_to_query<Binary>(Query &query, Predicate::Operator op, const PropertyExpression &expr)
+{
+    precondition(expr.indexes.empty(), "KeyPath queries not supported for data comparisons.");
+    Columns<Binary> column = expr.table_getter()->template column<Binary>(expr.col_ndx);
+    switch (op) {
+        case Predicate::Operator::NotEqual:
+            query.not_equal(expr.col_ndx, realm::null());
+            break;
+        case Predicate::Operator::Equal:
+            query.equal(expr.col_ndx, realm::null());
+            break;
+        default:
+            throw std::logic_error("Only 'equal' and 'not equal' operators supported when comparing against 'null'.");
+    }
+}
+
+template<>
+void do_add_null_comparison_to_query<Link>(Query &query, Predicate::Operator op, const PropertyExpression &expr)
+{
+    precondition(expr.indexes.empty(), "KeyPath queries not supported for object comparisons.");
+    switch (op) {
+        case Predicate::Operator::NotEqual:
+            query.Not();
+            REALM_FALLTHROUGH;
+        case Predicate::Operator::Equal:
+            query.and_query(query.get_table()->column<Link>(expr.col_ndx).is_null());
+            break;
+        default:
+            throw std::logic_error("Only 'equal' and 'not equal' operators supported for object comparison.");
+    }
+}
+
+void do_add_null_comparison_to_query(Query &query, Predicate::Comparison cmp, const PropertyExpression &expr)
+{
+    DataType type = expr.col_type;
+// FIXME: uncomment this and test it for LinkList and backlink column types
```

Comment:
What's the behavior with this code commented out? Does it give incorrect results, or do we end up throwing later on?

Reply:
Not sure, but I'll fix that.